python/dgl/backend/chainer/tensor.py

```diff
@@ -0,0 +1,536 @@
+        input = F.transpose(input, axes)
+        y = F.transpose(y, axes)
+
+    y = F.scatter_add(y, seg_id.data, input)
+
+    if dim != 0:
+        y = F.transpose(y, axes)
+
+    return y
+
+def unsorted_1d_segment_mean(input, seg_id, n_segs, dim):
+    n_ones = chainer.as_variable(
+        get_array_module(seg_id).ones_like(seg_id.data).astype(input.dtype))
+
+    w = unsorted_1d_segment_sum(n_ones, seg_id, n_segs, 0)
+    w = F.clip(w, 1., np.inf)
+    y = unsorted_1d_segment_sum(input, seg_id, n_segs, dim)
+
+    expand_dims = [1] * y.ndim
+    expand_dims[dim] = -1
+    y = y / F.reshape(w, expand_dims)
+
+    return y
+
+def boolean_mask(input, mask):
+    return input[mask.data]
+
+def equal(x, y):
+    return chainer.as_variable(x.data == y.data)
+
+def logical_not(input):
+    v = chainer.as_variable(
+        get_array_module(input).logical_not(input.data))
+    return v
+
+def unique(input):
+    v = chainer.as_variable(get_array_module(input).unique(input.data))
+    return v
+
+def full_1d(length, fill_value, dtype, ctx):
+    v = chainer.as_variable(
+        get_context_module(ctx).full((length,), fill_value, dtype=dtype))
+    return v
+
+def nonzero_1d(input):
+    v = chainer.as_variable(
+        get_array_module(input).nonzero(input.data)[0])
+    return v
+
+def sort_1d(input):
+    idx = get_array_module(input).argsort(input.data)
+    val = input[idx]
+    return val, idx
+
+def arange(start, stop):
+    return chainer.as_variable(np.arange(start, stop).astype('int64'))
+
+def rand_shuffle(arr):
+    idx = np.random.permutation(arr.shape[0])
+    return arr[idx]
+
+# zerocopy_to_dlpack and zerocopy_from_dlpack disabled
+
+def zerocopy_to_numpy(input):
+    # NOTE: zerocopy for CPU but not zerocopy for GPU
+    return asnumpy(input)
+
+def zerocopy_from_numpy(input):
+    return chainer.as_variable(input)
+
+def _zerocopy_to_dgl_ndarray(input_array):
+    if isinstance(input_array, np.ndarray):
+        return nd.array(np.ascontiguousarray(input_array))
+    elif isinstance(input_array, cupy.ndarray):
+        return nd.from_dlpack(cupy.ascontiguousarray(input_array).toDlpack())
+    else:
+        raise TypeError('Unknown type %s.' % type(input_array))
+
+def _zerocopy_from_dgl_ndarray(input):
+    if input.ctx.device_type == 1:      # cpu
+        return input.asnumpy()
+    elif input.ctx.device_type == 2:    # gpu
+        return cupy.fromDlpack(input.to_dlpack())
+    else:
+        raise TypeError('Unsupported device %d' % input.ctx.device_type)
+
+def zerocopy_to_dgl_ndarray(input):
+    return _zerocopy_to_dgl_ndarray(input.data)
+
+def zerocopy_from_dgl_ndarray(input):
+    return chainer.as_variable(_zerocopy_from_dgl_ndarray(input))
+
+
+# TODO(quan):
+# Currently I could not unify the NumPy-DLPack and CuPy-DLPack interaction
+# logic, so I resorted to old-style Chainer functions.
+# The new array backend ChainerX currently does not support DLPack.
+class BinaryReduce(chainer.Function):
```

Comment:
@ylfdq1118 I highly suggest we could further separate the logic into frame-specific and frame-agnostic part. For example, I feel what we need is an adaptor class `backend.Function` which unifies `torch.autograd.Function`, `mxnet.autograd.Function`, `chainer.Function`. By doing so, we could put this `BinaryReduce` class outside of backend, otherwise it's just too much of copy-pasting.